### PR TITLE
CDAP-14104 add ability to determine which ip to use for dataproc

### PR DESCRIPTION
**Pull request by albertshau**
Changed the Dataproc provisioner so that if it is running on GCP
in the same network as the dataproc cluster, it will by default
use the internal ip address when communicating with the cluster.
This makes it less likely that firewall rules will be an issue.
If users still want to use the external ip address, that is left
as an option.

Performed some refactoring so that CDAP no longer looks for the
'ip.external' node property, since that is no longer accurate.
Instead, added ip address as a required field in the Node class,
since it was functionally required anyway. This allows us to make
the contract explicit, instead of having required properties in
some Map.

Also made network and zone optional properties. When the provisioner
is running on GCP, the network and zone can be picked up from the
metadata server running on the VM.

### CODE REVIEW

**Comment by albertshau** [comment]
https://builds.cask.co/browse/CDAP-DUT6597

File: cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcConf.java [hunk 1]
```diff
@@ -213,15 +217,23 @@ private static DataProcConf create(Map<String, String> properties, @Nullable SSH
...
     return new DataProcConf(accountKey, region, zone, projectId, network,
                             masterNumNodes, masterCPUs, masterMemoryGB, masterDiskGB,
                             workerNumNodes, workerCPUs, workerMemoryGB, workerDiskGB,
-                            pollCreateDelay, pollCreateJitter, pollDeleteDelay, pollInterval, publicKey);
+                            pollCreateDelay, pollCreateJitter, pollDeleteDelay, pollInterval,
+                            preferExternalIP, publicKey);
   }
 
-  private static String getString(Map<String, String> properties, String key, String defaultVal) {
+  // the UI never sends nulls, it only sends empty strings.
```
Comment (anwar6953) [inline cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcConf.java:229]: "UI never sends nulls"
I think this is true for some widget types, such as text widget type.
But its not true for all widget types, such as selector widgets.

Comment (albertshau) [reply cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcConf.java:229]: selectors don't have a way to not select anything (at least in the provisioner page)

**Review by chtyim — APPROVED** [review]
LGTM

File: cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcConf.java (hunk 1 above)
Comment (anwar6953) [reply cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcConf.java:229]: oh ok; my bad. I must have misunderstood something.

**Pull request merged**